File: Lab 2/lru.py
# LRU
'''
Let capacity be the number of pages that
memory can hold.  Let set be the current
set of pages in memory.

1- Start traversing the pages.
 i) If set holds less pages than capacity.
   a) Insert page into the set one by one until 
      the size  of set reaches capacity or all
      page requests are processed.
   b) Simultaneously maintain the recent occurred
      index of each page in a map called indexes.
   c) Increment page fault
 ii) Else 
   If current page is present in set, do nothing.
   Else 
     a) Find the page in the set that was least 
     recently used. We find it using index array.
     We basically need to replace the page with
     minimum index.
     b) Replace the found page with current page.
     c) Increment page faults.
     d) Update index of current page.

2. Return page faults.
'''
def page_faults(pages, frame_capacity):
   page_fault = 0
   frame = []
   indexes = []
   least_recent_index = 0
   for i in range(len(pages)):
      if len(frame) != frame_capacity:
         frame.append(pages[i])
         page_fault += 1
         indexes.append(i)
      else:
         if pages[i] in frame:
            # update index
            indexes[frame.index(pages[i])] = i
         else:
            frame[indexes.index(least_recent_index)] = pages[i]
            indexes[indexes.index(least_recent_index)] = i
            page_fault += 1

         least_recent_index = min(indexes)   
            
   return page_fault

pages = [7,0,1,2,0,3,0,4,2,3,0,3,2]
print("Page Fault(s): ", page_faults(pages, 4))

File: Lab 2/lru.java
import java.util.LinkedList;
import java.util.ArrayList;
import java.util.Collections;

public class lru {
    public static void main(String[] args){
        LinkedList<Integer> l = new LinkedList<>();
        int[] k = {7,0,1,2,0,3,0,4,2,3,0,3,2};
        int frame_capacity = 4;

        lru(k, l, frame_capacity);
    }

    public static void lru(int[] arr,LinkedList<Integer> link,int capacity){
        
        // int[] frame = new int[0];
        int page_fault = 0;
        int len = 0;
        ArrayList<Integer> index = new ArrayList<>();
        int lru_index = 0;

        for(int i=0;i<arr.length;i++){
            if(len < capacity){
                link.add(arr[i]);
                page_fault++;
                index.add(i);
                // System.out.println("masuk");
            }
            else{
                // System.out.println("Penuh");
                if(link.contains(arr[i])){
                    index.set(link.indexOf(arr[i]),i);
                    // System.out.println(index.get(arr[i]));
                    // System.out.println("yang ada");
                }
                else{
                    link.set(index.indexOf(lru_index),arr[i]);
                    index.set(index.indexOf(lru_index),i);
                    page_fault++;
                    // System.out.println("yang takde");
                }
                lru_index = Collections.min(index);
                // System.out.println("lru "+lru_index);
            }
            len++;
            // System.out.println();

            
        }

        System.out.println("The total page fault is: "+ page_fault);
        System.out.println("The final page is: "+link);
        // System.out.println(index);
    }
}
